### PR TITLE
Master payment paypal apocalypse aho

### PR DESCRIPTION
**Pull request by AdrienHorgnies**
Adapt payment_paypal module to new payment API.

Task: https://www.odoo.com/web#id=2333036&action=333&active_id=1251&model=project.task&view_type=form&cids=1&menu_id=4720

### CODE REVIEW

File: addons/payment_paypal/models/payment_acquirer.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    paypal_seller_account = fields.Char(
+        'Merchant Account ID', groups='base.group_user',
+        help='The Merchant ID is used to ensure communications coming from Paypal are valid and secured.')
+    paypal_use_ipn = fields.Boolean('Use IPN', default=True, help='Paypal Instant Payment Notification',
+                                    groups='base.group_user')
+    paypal_pdt_token = fields.Char(string='PDT Identity Token',
+                                   help='Payment Data Transfer allows you to receive notification of successful payments as they are made.',
+                                   groups='base.group_user')
+
+    def _compute_fees(self, amount, currency_id, country_id):
```
Comment (AntoineVDV) [inline addons/payment_paypal/models/payment_acquirer.py]: self note
```suggestion
    def _compute_fees(self, amount, currency, country):
```

File: addons/payment_paypal/models/payment_acquirer.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        selection_add=[('paypal', "Paypal")], ondelete={'paypal': 'set default'})
+    paypal_email_account = fields.Char(
+        string="Email", required_if_provider='paypal', groups='base.group_system')
+    paypal_seller_account = fields.Char(
+        string="Merchant Account ID", groups='base.group_system',
+        help="This ensures that communications coming from Paypal are valid and secured.")
+    paypal_pdt_token = fields.Char(
+        string="PDT Identity Token",
+        help="This allows to receive notifications of successful payments as they are made.",
+        groups='base.group_system')
```
Comment (Feyensv) [inline addons/payment_paypal/models/payment_acquirer.py:24]: move after the string to gain a line :D

File: addons/payment_paypal/controllers/main.py [hunk 3]
```diff
@@ -1,119 +1,95 @@
...
+            and only have as params the transaction ID and the PDT Identity Token (under the key
+            `at`, as per previous versions of this method).
+          - The payment data should be parsed from the response of the check request.
+        In practice, however, the transaction ID is never given by PayPal and the documentation
+        has no mention of `_notify_synch` nor `at`. Because of this, PDT cannot be verified as
+        prescribed by the documentation.
+        Nevertheless, previous versions of this method used a bad heuristic (assessing the presence
+        of the optional, PDT-specific, param `amt`) to determine whether the notification was a PDT.
+        Since PDT notifications have in practice always been successfully authenticated by using the
+        IPN protocol, this method does explicitly that for both PDT and IPN.
```
Comment (Feyensv) [inline addons/payment_paypal/controllers/main.py:66]: Perfect explanation :+1: :100:

File: addons/payment_paypal/models/payment_transaction.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            **processing_values,
+            'address1': self.partner_address,
+            'business': self.acquirer_id.paypal_email_account,
+            'city': self.partner_city,
+            'cmd': '_xclick',
+            'country': self.partner_country_id.code,
+            'currency_code': self.currency_id.name,
+            'email': self.partner_email,
+            'first_name': partner_first_name,
+            'handling': self.acquirer_id.fees_active and self.fees,
```
Comment (Feyensv) [inline addons/payment_paypal/models/payment_transaction.py]: I'm not sure we should check at this step if the acquirer has fees enabled.  If the transaction has fees, it means that the acquirer had fees enabled when the transaction was created, and we have to show it, even if the configuration of the acquirer changed no?

File: addons/payment_paypal/models/payment_transaction.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        return tx
+
+    def _process_feedback_data(self, data):
+        self.ensure_one()
+        if self.acquirer_id.provider != 'paypal':
+            super()._process_feedback_data(data)
+
+        txn_id = data.get('txn_id')
+        txn_type = data.get('txn_type')
+        if not any((txn_id, txn_type)):
```
Comment (Feyensv) [inline addons/payment_paypal/models/payment_transaction.py]: If only one of the two is missing, you won't raise here.

File: addons/payment_paypal/models/payment_transaction.py (hunk 4 above)
Comment (AntoineVDV) [reply addons/payment_paypal/models/payment_transaction.py]: ~~Indeed.~~
Actually, you are getting things wrong but still made a point^^ This is code from `_get_specific_rendering_values` which means that we are in the same flow step as that of the transaction creation. There is no way that the configuration of the acquirer could have changed in the meantime. However, it's true that we don't need to check if fees are actives before setting the value in the dict.

**Pull request merged**